README.md
# log

[![Greenkeeper badge](https://badges.greenkeeper.io/Codeshare/log.svg)](https://greenkeeper.io/)

Codeshare Application Logger

# Usage
```js
import log from '@codeshare/log'

log.fatal('message', {/*...*/})
log.error('message', {/*...*/})
log.warn('message', {/*...*/})
log.info('message', {/*...*/})
log.debug('message', {/*...*/})
log.trace('message', {/*...*/})
```

```js
log.logRequest({/*...*/})
```

```js
await log.onFinish()
```


CHANGELOG.md
# 4.0.0

- major: rewrite in ts and pkg upgrades


src/__tests__/index.test.ts
describe('@codeshare/log', () => {
  let log

  beforeEach(async () => {
    process.env.APP_LOG_LEVEL = 'trace'
    process.env.APP_NAME = 'test'
    const module = await import('../index')
    log = module.default
  })

  it('should log', () => {
    log.info('yo')
  })
})


src/index.ts
import BaseError from 'baseerr'
import Bunyan from 'bunyan'
import { LoggingBunyan } from '@google-cloud/logging-bunyan'
import { Writable } from 'stream'
import errToJSON from 'error-to-json'
import { get } from '@codeshare/env'
import reqToJSON from 'request-to-json'
import sparkToJSON from 'spark-to-json'

let keepAliveTimer: NodeJS.Timer | undefined
function keepAlive(check: Function) {
  keepAliveTimer = setTimeout(function () {
    if (!check()) keepAlive(check)
  }, 1000)
}

class LoggerError extends BaseError<{}> {}

enum levels {
  'fatal' = 'fatal',
  'error' = 'error',
  'warn' = 'warn',
  'info' = 'info',
  'debug' = 'debug',
  'trace' = 'trace',
}

const level = get('APP_LOG_LEVEL')
  .default('fatal')
  .asEnum<levels>(Object.values(levels))

class Logger extends Bunyan {
  // @ts-ignore
  streams: Array<{ stream: Writable }>
  logEnded = false

  constructor() {
    super({
      name: get('APP_NAME').required().asString(),
      level,
      serializers: {
        ctx: ctxToJSON,
        err: errToJSON,
        req: reqToJSON,
        proxyReq: reqToJSON,
        spark: sparkToJSON,
      },
      streams: get('ENABLE_GCLOUD_LOG').asBool()
        ? // log to google-cloud
          [
            new LoggingBunyan({
              logName: process.env.APP_NAME,
              serviceContext: {
                service: process.env.APP_NAME,
                version: process.env.APP_VERSION,
              },
            }).stream(level),
            {
              level: 'fatal',
              stream: process.stdout,
            },
          ]
        : // log to stdout
          [
            {
              stream: process.stdout,
            },
          ],
    })
    this.on('error', (err: Error) =>
      console.error(LoggerError.wrap(err, 'LOG ERROR!')),
    )
  }

  // @ts-ignore
  fatal(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: fatal:', msg, data)
    super.fatal(data || {}, msg)
  }
  // @ts-ignore
  error(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: error:', msg, data)
    super.error(data || {}, msg)
  }
  // @ts-ignore
  warn(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: warn:', msg, data)
    super.warn(data || {}, msg)
  }
  // @ts-ignore
  info(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: info:', msg, data)
    super.info(data || {}, msg)
  }
  // @ts-ignore
  debug(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: debug:', msg, data)
    super.debug(data || {}, msg)
  }
  // @ts-ignore
  trace(msg: string, data?: {}): void {
    if (this.logEnded) console.log('STREAM ENDED BEFORE: trace:', msg, data)
    super.trace(data || {}, msg)
  }

  async end() {
    this.logEnded = true
    const unFinishedStreams: Array<Writable> = []
    this.streams.forEach((s) => {
      if (!s.stream) return
      if (s.stream === process.stdout) return
      if (s.stream === process.stderr) return
      unFinishedStreams.push(s.stream)
    })
    await Promise.all(
      unFinishedStreams.map(function (stream) {
        return new Promise(function (resolve) {
          if (stream.writableFinished) return resolve()
          keepAlive(function () {
            return stream.writableFinished
          })
          stream.on('finish', function () {
            if (keepAliveTimer != null) clearTimeout(keepAliveTimer)
            resolve()
          })
          stream.end()
        })
      }),
    )
  }
}

export default new Logger()

type Ctx = {
  token?: string
  me?: {}
  session?: {
    token?: string
    user?: {}
  }
}
function ctxToJSON({ token, me, session }: Ctx) {
  const json: Ctx = {
    token,
    me,
  }
  if (session) {
    json.session = {
      token: session.token,
      user: session.user,
    }
  }
  return json
}


lib/gcloud-stream.js
'use strict'

const env = require('@codeshare/env')
env.assert([
  'NODE_ENV',
  'APP_NAME',
  'APP_VERSION'
])

const bunyan = require('bunyan')
const client = require('@google/cloud-errors').start({
  reportUncaughtExceptions: false,
  serviceContext: {
    service: process.env.APP_NAME,
    version: process.env.APP_VERSION
  }
})
const errToJSON = require('error-to-json')
const kp = require('keypather')()
const parseErr = require('error-to-json').parse
const ReqContainer = require('@google/cloud-errors/lib/classes/request-information-container.js')
const through2 = require('through2')

const levelToName = {
  60: 'fatal',
  50: 'error',
  40: 'warn',
  30: 'info',
  20: 'debug',
  10: 'trace'
}

module.exports = {
  type: 'raw',
  level: bunyan.ERROR,
  stream: through2.obj(function (logObj, enc, cb) {
    if (logObj.err && !kp.get(logObj, 'err.data.dontReport')) {
      const ctx = logObj.ctx || {}
      const userId = kp.get(ctx, 'me.id')
      let reqContainer
      if (ctx.req) {
        reqContainer = new RequestContainer()
        reqContainer.setMethod(ctx.req.method)
        reqContainer.setUrl(ctx.req.url)
        reqContainer.setUserAgent(ctx.req.headers['user-agent'])
        reqContainer.setReferrer(ctx.req.headers['referrer'] || ctx.req.headers['referer'])
        if (ctx.req.socket && ctx.req.socket.remoteAddress) {
          reqContainer.setRemoteAddress(ctx.req.socket.remoteAddress)
        }
      } else if (ctx.spark && ctx.rootSpanData) {
        reqContainer = new RequestContainer()
        reqContainer.setMethod('GET')
        reqContainer.setUrl(kp.get(ctx.rootSpanData, 'spanData.name'))
        reqContainer.setRemoteAddress(ctx.spark.headers['x-forwarded-for'] || ctx.spark.remote)
      }
      if (reqContainer) {
        reqContainer.setStatusCode(logObj.err.statusCode || 500)
      }
      const err = parseErr(logObj.err)
      err.stack += '---\nlog.error: ' + logObj.msg
      client.report(err, reqContainer, null, userId, cb)
    } else {
      cb()
    }
  })
}


lib/sentry-stream.js
'use strict'

const env = require('@codeshare/env')
env.assert([
  'NODE_ENV',
  'APP_GIT_SHA',
  'APP_VERSION'
])

const bunyan = require('bunyan')
const errToJSON = require('error-to-json')
const kp = require('keypather')()
const raven = require('raven')
const through2 = require('through2')

const levelToName = {
  60: 'fatal',
  50: 'error',
  40: 'warn',
  30: 'info',
  20: 'debug',
  10: 'trace'
}

module.exports = {
  type: 'raw',
  level: bunyan.ERROR,
  stream: through2.obj(function (logObj, enc, cb) {
    if (logObj.err && !kp.get(logObj, 'err.data.dontReport')) {
      const client = new raven.Client(process.env.SENTRY_DSN, {
        environment: process.env.NODE_ENV,
        release: process.env.APP_GIT_SHA,
        level: levelToName[logObj.level],
        extra: logObj
      })
      // set tags
      client.setTagsContext({
        pid: process.pid,
        version: process.env.APP_VERSION
      })
      // add user info, if it exists
      if (logObj.ctx && logObj.ctx.me) {
        const userInfo = pick(logObj.ctx.me, ['id', 'email'])
        client.setUserContext(userInfo)
      }
      // send error to sentry
      const err = errToJSON.parse(logObj.err)
      client.captureException(err)
    }
    cb()
  })
}


index.js
'use strict'

require('@codeshare/env').assert([
  'APP_LOG_LEVEL',
  'APP_NAME',
  'ENABLE_GCLOUD_LOG',
])

const AppError = require('codeshare-error')
const bunyan = require('bunyan')
const errToJSON = require('error-to-json')
const { LoggingBunyan } = require('@google-cloud/logging-bunyan')
const pick = require('101/pick')
const reqToJSON = require('request-to-json')
const shimmer = require('shimmer')
const sparkToJSON = require('spark-to-json')

// "fatal" (60): The service/app is going to stop or become unusable now. An operator should definitely look into this soon.
// "error" (50): Fatal for a particular request, but the service/app continues servicing other requests. An operator should look at this soon(ish).
// "warn"  (40): A note on something that should probably be looked at by an operator eventually.
// "info"  (30): Detail on regular operation.
// "debug" (20): Anything else, i.e. too verbose to be included in "info" level.
// "trace" (10): Logging from external libraries used by your app or very detailed application logging.
const LOG_LEVEL = process.env.APP_LOG_LEVEL || 'fatal'
// hardcode for now..
const TRACE_CONTEXT_HEADER_NAME = 'x-cloud-trace-context'

const log = (module.exports = bunyan.createLogger({
  name: process.env.APP_NAME,
  level: LOG_LEVEL,
  serializers: {
    ctx: pick(['token', 'me', 'session.token', 'session.user']),
    err: errToJSON,
    req: reqToJSON,
    proxyReq: reqToJSON,
    spark: sparkToJSON,
  },
  streams: process.env.ENABLE_GCLOUD_LOG
    ? // log to google-cloud
      [
        new LoggingBunyan({
          logName: process.env.APP_NAME,
          serviceContext: {
            service: process.env.APP_NAME,
            version: process.env.APP_VERSION,
          },
        }).stream(),
        {
          level: 'fatal',
          stream: process.stdout,
        },
      ]
    : // log to stdout
      [
        {
          stream: process.stdout,
        },
      ],
}))

log.on('error', function (err) {
  console.error(new AppError(500, 'LOG ERROR!', err))
})

let logEnded = false

const names = ['fatal', 'error', 'warn', 'info', 'debug', 'trace']
names.forEach(function (name) {
  shimmer.wrap(log, name, function (orig) {
    return function (obj, msg) {
      if (logEnded) console.log('STREAM ENDED BEFORE:', msg)
      // allows for args in any order..
      let tmp
      if (obj && typeof obj === 'string') {
        // args are reversed.. swap them
        tmp = msg
        msg = obj
        obj = tmp
      }
      obj = obj || {}
      // default log values
      Object.assign(obj, {
        env: process.env.NODE_ENV,
        commit: process.env.APP_GIT_COMMIT,
        version: process.env.APP_VERSION,
      })
      // pick req off ctx if it exists
      if (obj.ctx && obj.ctx.req && !obj.req) {
        obj.req = obj.ctx.req
      }
      // pick trace key off of header
      const traceKey =
        obj.req && obj.req.headers && obj.req.headers[TRACE_CONTEXT_HEADER_NAME]
      if (traceKey) {
        obj[LoggingBunyan.LOGGING_TRACE_KEY] = traceKey
      }
      return orig.call(this, obj, msg)
    }
  })
})

log.logRequest = function (obj) {
  obj = obj || {}
  // pick req off ctx if it exists
  if (obj.ctx && obj.ctx.req && !obj.req) {
    obj.req = obj.ctx.req
  }
  // stackdriver http request logging
  if (obj.req) {
    const ctx = obj.ctx || {}
    const req = obj.req
    const res = ctx.res || {}
    const headers = req.headers || {}
    const connection = req.connection || {}
    const socket = req.socket || connection.socket || {}
    const remoteIp =
      headers['x-forwarded-for'] ||
      connection.remoteAddress ||
      socket.remoteAddress
    const httpRequest = {
      requestMethod: req.method || ctx.method,
      requestUrl: req.url || ctx.url,
      // requestSize: int64,
      status: res.statusCode || ctx.status,
      // responseSize: int64,
      remoteIp: remoteIp,
      serverIp: socket.localAddress,
      referer: headers['referer'],
      userAgent: headers['user-agent'],
      // latency: Duration,
      // cacheLookup: bool,
      // cacheHit: bool,
      // cacheValidatedWithOriginServer: bool,
      // cacheFillBytes: int64,
      protocol: 'HTTP/1.1',
    }
    obj.httpRequest = obj.httpRequest
      ? Object.assign(httpRequest, obj.httpRequest)
      : httpRequest
  }
  log.info('http request', obj)
}

let keepAliveTimer = null
function keepAlive(check) {
  keepAliveTimer = setTimeout(function () {
    if (!check()) keepAlive(check)
  }, 1000)
}

log.end = function () {
  const unFinishedStreams = []
  log.streams.forEach((s) => {
    if (!s.stream) return
    if (s.stream === process.stdout) return
    if (s.stream === process.stderr) return
    unFinishedStreams.push(s.stream)
  })
  return Promise.all(
    unFinishedStreams.map(function (stream) {
      return new Promise(function (resolve) {
        if (stream.finished) return resolve()
        keepAlive(function () {
          return stream.finished
        })
        stream.on('finish', function () {
          clearTimeout(keepAliveTimer)
          resolve()
        })
        logEnded = true
        stream.end()
      })
    }),
  )
}
